# 4.5 DATA CONSISTENCY CHECKS
****

**SCRIPT CONTENTS:**

1. Importing Libraries & Files
2. Data Consistency Checks
      - Mixed-Type Data
      - MIssing Values
      - Duplicates
3. Tyding Up DataFrame
4. Exporting Clean DataFrame: **departments_checked.csv & orders_checked.csv**

# Importing Libraries

In [2]:
#Importing libraries
import pandas as pd
import numpy as np
import os

In [6]:
path= r'D:\Instacart Basket Analysis'

In [7]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [8]:
df_ords = pd.read_csv(os.path.join(path, '02 Data',  'Prepared Data','orders_wrangled.csv'), index_col = False)

In [9]:
df_ords. describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

In [10]:
#create a dataframe
df_test = pd.DataFrame()

In [11]:
#create a mixed type column
df_test['mix'] = ['a','b',1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
df_prods_clean =df_prods[df_prods['product_name'].isnull()== False]

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
df_prods.dropna(inplace = True)

In [22]:
df_prods.dropna(subset = ['product_name'], inplace = True)

In [23]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [24]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [28]:
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [29]:
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

#days_since_prior_order has 206209 missing values. Since the first entry of user_id = 1 and 2 had a NaN value for 'days_since_prior_order' I theorized that the first transaction of every user would have a NaN value for 'days_since_prior_order'. If the largest user_id is equal to the number of missing values, then the theory is likely true. The max value for user_id = 206,209 (the same as the number of missing values)

In [30]:
df_ords.days_since_prior_order.describe()

count    3.214874e+06
mean     1.111484e+01
std      9.206737e+00
min      0.000000e+00
25%      4.000000e+00
50%      7.000000e+00
75%      1.500000e+01
max      3.000000e+01
Name: days_since_prior_order, dtype: float64

In [31]:
df_ords['days_since_prior_order'].value_counts()

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [32]:
df_ords['days_since_prior_order'].fillna(0, inplace=True)

I imputed a flag value of '0' for all NaN entries in 'days_since_prior_order' so it would not affect analysis results regarding purchase frequency. The high frequency of '30' values (likely due a default value entered at the end of the month, or from customers with monthly recurring orders) stretched the mean, making it a poor imputation. '7' is the median is already the second most frequent value - 30% larger than the third most frequent. Adding all NaN values to this could skew analysis results.

In [33]:
df_ords.shape

(3421083, 7)

In [34]:
df_ords_clean =df_ords[df_ords['days_since_prior_order'].isnull()== False]

In [35]:
df_ords_clean.shape

(3421083, 7)

In [36]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [37]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [38]:
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [39]:
df_ords_clean_no_dups.shape

(3421083, 7)

In [36]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [37]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_checked.csv'))